### PR TITLE
[12.0][ADD] sale_stock_last_date: New module. Display last delivery date in sale order lines

### PR DESCRIPTION
**Pull request by sergio-teruel**
cc @Tecnativa TT31039

ping @carlosdauden @chienandalu

### CODE REVIEW

**Review by carlosdauden — APPROVED** [review]

File: sale_stock_last_date/tests/test_sale_stock_last_date.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    def test_last_delivery_date(self):
+        self.order.action_confirm()
+        self.assertFalse(self.order.order_line.last_delivery_date)
+        # Partial delivery one
+        picking = self.order.picking_ids
+        picking.action_assign()
+        picking.move_lines.quantity_done = 2.0
+        picking.action_done()
+        self.assertEqual(
+            self.order.order_line.last_delivery_date, picking.date)
```
Comment (carlosdauden) [inline sale_stock_last_date/tests/test_sale_stock_last_date.py]: Travis singleton error at this line for several sale order lines :confused:

**Review by bodedra — COMMENTED** [review]
Functional reviewed. :+1: 

Tested with regular order and backorder. Set correct last delivery date in SO line.

Also verified Sales Analysis report. Looks good.

Nice addition.

Minor: Please fix test case.

**Review by chienandalu — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Travis is still red

File: sale_stock_last_date/models/sale.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+from odoo import api, fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    last_delivery_date = fields.Datetime(
+        string="Last delivery date",
+        compute="_compute_last_delivery_date",
+        store=True
```
Comment (chienandalu) [inline sale_stock_last_date/models/sale.py:12]: In a large DB, this is going to have a hard time computing on the init? Maybe we could add a init hook to ease it?

File: sale_stock_last_date/tests/test_sale_stock_last_date.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.assertFalse(self.order.order_line.last_delivery_date)
+        # Partial delivery one
+        picking = self.order.picking_ids
+        picking.action_assign()
+        picking.move_lines.quantity_done = 2.0
+        picking.action_done()
+        # Discard any so line like as delivery line for tests in travis
+        so_line = self.order.order_line.filtered(
+            lambda ln: ln.product_id == self.product)
+        self.assertEqual(so_line.last_delivery_date, picking.date)
```
Comment (chienandalu) [inline sale_stock_last_date/tests/test_sale_stock_last_date.py]: The move date and picking date fell into different seconds: https://app.travis-ci.com/github/OCA/sale-workflow/jobs/526378588#L2564

Comment (chienandalu) [reply sale_stock_last_date/tests/test_sale_stock_last_date.py]: Maybe an approach could be using freegun and comparing just the date part

**Comment by sergio-teruel** [comment]
Changes done

**Review by chienandalu — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

File: sale_stock_last_date/models/sale.py (hunk 2 above)
Comment (pedrobaeza) [reply sale_stock_last_date/models/sale.py:12]: Oh, this one is still here... You should pre-create the column and leave it empty for all ones IMO.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: sale_stock_last_date/models/sale.py (hunk 2 above)
Comment (pedrobaeza) [reply sale_stock_last_date/models/sale.py:12]: Or do an equivalent SQL for doing all at once.

**Comment by sergio-teruel** [comment]
@bodedra Test fixed

**Comment by bodedra** [comment]
> @bodedra Test fixed

Thanks @sergio-teruel 

Would you please address @pedrobaeza 's suggested changes? Afterwards,  I will do the final functional review again. Thank you.

**Comment by sergio-teruel** [comment]
Changes done!!!

**Review by pedrobaeza — APPROVED** [review]

**Comment by rousseldenis** [comment]
@sergio-teruel I would have named it differently as from module name, we don't know 'which' date it is.

**Comment by sergio-teruel** [comment]
@rousseldenis Sorry the module has already been merged....

**Comment by rousseldenis** [comment]
> @rousseldenis Sorry the module has already been merged....

Yes, I know. It was a constructive remark for records and maybe for future migration :smiley:

**Comment by pedrobaeza** [comment]
I don't see it bad at all, being a technical name. `sale_stock` is saying that is about stock part of sale (the delivery), and `last_date` indicates that it's the last. `sale_stock_last_delivery_date` may be more accurate, but it introduces also the taboo `delivery` word that mixes up with all the carrier stuff.

**Comment by rousseldenis** [comment]
> I don't see it bad at all, being a technical name. `sale_stock` is saying that is about stock part of sale (the delivery), and `last_date` indicates that it's the last. `sale_stock_last_delivery_date` may be more accurate, but it introduces also the taboo `delivery` word that mixes up with all the carrier stuff.

IMHO pointing Odoo module names is a bad example :smiley:

I did this in order to help people finding what they look for. For me, nothing is taboo. Naming explicitly things is not bad.. as in this case, it's exactly what the module does

**Comment by pedrobaeza** [comment]
The naming scheme for knowing the dependencies is important. If Odoo has a bad name, we can't know reinvent a new family naming scheme due to this.

**Pull request merged**